### PR TITLE
Backport bitcoin#15773: test: Add BitcoinTestFramework::sync_* methods (incomplete backport)

### PR DESCRIPTION
**Pull request by codablock**
This backports bitcoin#15773 and also moves the `bump_mocktime` call into the new TestFramework.sync_mempool method, instead of doing this for every `sync_mempool` call. This is needed as it turned out that much more places need this logic as otherwise tests fail randomly.

Backport is done commit by commit and scripted diffs are not really backported but instead re-applied to ensure we don't miss any Dash specific places.

I left out fafe5f0d09db6778173a1079d93c314e9317708c, as it had too many conflicts. Better to do this later by redoing the lint checks and following the suggestions.

### CODE REVIEW

File: test/functional/test_framework/test_framework.py [hunk 1]
```diff
@@ -347,9 +347,15 @@ def join_network(self):
...
+        sync_blocks(nodes or self.nodes, **kwargs)
+
+    def sync_mempools(self, nodes=None, **kwargs):
+        sync_mempools(nodes or self.nodes, **kwargs)
+
     def sync_all(self, nodes=None, **kwargs):
-            sync_blocks(nodes or self.nodes, **kwargs)
-            sync_mempools(nodes or self.nodes, wait=0.1, wait_func=lambda: self.bump_mocktime(3, True), **kwargs)
+        self.sync_blocks(nodes, **kwargs)
+        self.sync_mempools(nodes or self.nodes, wait=0.1, wait_func=lambda: self.bump_mocktime(3, True), **kwargs)
```
Comment (PastaPastaPasta) [inline test/functional/test_framework/test_framework.py]: Why is this sync mempool call different https://github.com/bitcoin/bitcoin/pull/15773/commits/fa6dc7c5c3477bda04481f080d430f63539974e2#diff-64721c5ee64d44f7114d6d0d2226db4dR417?

Should it not be... 
```suggestion
        self.sync_mempools(nodes, wait=0.1, wait_func=lambda: self.bump_mocktime(3, True), **kwargs)
```

File: test/functional/test_framework/test_framework.py [hunk 2]
```diff
@@ -451,7 +457,7 @@ def _initialize_chain(self, extra_args=None, stderr=None):
                         self.nodes[peer].generate(1)
                         block_time += 156
                     # Must sync before next peer starts generating blocks
-                    sync_blocks(self.nodes)
+                    self.sync_blocks(self.nodes)
```
Comment (PastaPastaPasta) [inline test/functional/test_framework/test_framework.py:466]: noting difference https://github.com/bitcoin/bitcoin/pull/15773/commits/fa6dc7c5c3477bda04481f080d430f63539974e2#diff-64721c5ee64d44f7114d6d0d2226db4dR503

```suggestion
                    self.sync_blocks()
```

File: test/functional/test_framework/test_framework.py (hunk 1 above)
Comment (codablock) [reply test/functional/test_framework/test_framework.py]: Yepp, you're right. Fixed it.

File: test/functional/test_framework/test_framework.py (hunk 2 above)
Comment (codablock) [reply test/functional/test_framework/test_framework.py]: Fixed/squashed

**Comment by codablock** [comment]
@Pasta regarding missing https://github.com/bitcoin/bitcoin/commit/fafe5f0d09db6778173a1079d93c314e9317708c, please see the PR description (I added the info about fafe5f0d09db6778173a1079d93c314e9317708c a few minutes after creating the PR)

**Comment by codablock** [comment]
fyi, tests will verify likely not get green with this PR as a lot of other failures are fixed with #3410

**Comment by PastaPastaPasta** [comment]
> 
> 
> @Pasta regarding missing [bitcoin@fafe5f0](https://github.com/bitcoin/bitcoin/commit/fafe5f0d09db6778173a1079d93c314e9317708c), please see the PR description (I added the info about [fafe5f0](https://github.com/dashpay/dash/commit/fafe5f0d09db6778173a1079d93c314e9317708c) a few minutes after creating the PR)

Whoops! I guess it is a good thing to read the whole PR description! In that case can we update the PR title to say "partial"?

**Comment by codablock** [comment]
@PastaPastaPasta I added (incomplete backport) to the PR description

**Comment by codablock** [comment]
Tests in #3410 just got green, so it should be safe to merge this PR and then proceed with merging #3410

**Pull request merged**